Drivers: Ethernet: ch32v30x SoC

BOJIT opened the pull request:
This Draft PR adds Ethernet support for the CH32V307 SOC.

This has been fully tested on the CH32V307 dev board with the internal 10M PHY. Support for external PHY options has been added but I don't have the hardware to test it.

This currently depends on addition of a custom `arch_cpu_idle` implementation to fix for a hardware bug. This cannot be merged until the approach for this has been decided (https://github.com/zephyrproject-rtos/zephyr/pull/101391)

BOJIT committed:
drivers: rng: Add RNG Peripheral driver

This adds RNG support for the CH32V20x_30x family.
Driver required for network stack support.

Signed-off-by: James Bennion-Pedley <james@bojit.org>

pdgendt commented on drivers/ethernet/Kconfig.wch:
```diff
@@ -0,0 +1,51 @@
+
+config ETH_WCH_MULTICAST_FILTER
+	bool "Multicast hash filter support"
+	default n
```
Ditto

pdgendt commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,904 @@
+#include "eth.h"
+
+/* Internal 10BASE-T PHY 50R*4 pull-up resistance enable or disable */
+#define ETH_Internal_Pull_Up_Res_Enable  ((uint32_t)0x00100000)
```
Bit flags are typically set using the `BIT` macro, and defines use uppercase.

```c
#define ETH_INTERNAL_PULL_UP_RES_ENABLE BIT(20)
```

pdgendt commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,904 @@
+		.clk_tx_id = DT_INST_CLOCKS_CELL_BY_IDX(inst, 0, id),                              \
+		.clk_rx_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR_BY_IDX(inst, 1)),                  \
+		.clk_rx_id = DT_INST_CLOCKS_CELL_BY_IDX(inst, 1, id),                              \
+		.use_random_mac = DT_INST_PROP(inst, zephyr_random_mac_address),                   \
```
Use the MAC address configuration as described here: https://docs.zephyrproject.org/latest/connectivity/networking/api/mac_config.html

pdgendt commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,904 @@
+	const struct eth_wch_config *config = dev->config;
+	clock_control_subsys_t clock_sys;
+
+	int ret = 0;
```
No need to init as it's assigned first.

pdgendt commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,904 @@
+	clock_sys = (clock_control_subsys_t *)(uintptr_t)config->clk_rx_id;
+	ret |= clock_control_on(config->clk_rx_dev, clock_sys);
+
+	if (ret) {
```
```c
if (ret != 0) {
```

maass-hamburg commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,904 @@
+		eth->MACA0LR = (data->mac_addr[3] << 24) | (data->mac_addr[2] << 16) |
+			       (data->mac_addr[1] << 8) | data->mac_addr[0];
+		net_if_set_link_addr(data->iface, data->mac_addr, sizeof(data->mac_addr),
+				     NET_LINK_ETHERNET);
```
move into seperate function and use here and in iface init

BOJIT commented:
> > As for the H417, I can't find the reference manual
> 
> Only in chinese last i checked, i did the work for it on ch32fun side but not even started porting it to zephyr.
> 
> > Yep, it should work, the register maps are the same.
> 
> Great 👍 !

@VynDragon sorry, had another look at the reference manual and it turns out that the V203/208 Ethernet peripheral is completely different register maps, so won't be compatible :(. This driver will only support the V307, V317 and F207.

BOJIT commented:
> You need one of the zephyr badge for the ci to run without someone manually triggering it

Ah, I see. Is that this one? [Claim Your Zephyr Badge](https://forms.gle/oCw9iAPLhUsHTapc8)? Or the [nomination template](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=Role+Nomination&template=006_nomination.yml)?

I have merged a handful of patches in the past.

VynDragon commented:
No it's the github one, possibly 'contributor' badge you get after one merged PR (if you merged one and didnt get it you probably should ask for it), and also automatically if you are put in MAINTAINERS.yml

VynDragon commented:
I asked on discord and they said it should've been triggering... Probably will fix itself in the next PR or something. Either way contirbutor badge is good to have.

maass-hamburg commented on drivers/ethernet/Kconfig.wch:
```diff
@@ -0,0 +1,49 @@
+
+config ETH_WCH_MULTICAST_FILTER
+	bool "Multicast hash filter support"
+	imply CRC
```
```suggestion
	select CRC
```
looks like it is needed, therefore it has to be slected

maass-hamburg commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,786 @@
+struct eth_wch_data {
+	struct net_if *iface;
+	uint8_t mac_addr[NET_ETH_ADDR_LEN];
+	struct k_mutex tx_mutex;
```
tx mutex is not needed, the net subsystem already provides one. That's why it also was removed from the other ethernet drivers.

maass-hamburg commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,786 @@
+		is_first_init = true;
+	}
+
+	if (is_first_init) {
```
combine this if with the if before, not need for this extra bool var.

maass-hamburg commented on drivers/ethernet/Kconfig.wch:
```diff
@@ -0,0 +1,49 @@
+	  Preemptive scheduling can lead to more responsive handling of network traffic,
+	  especially under high load.
+
+config ETH_WCH_HW_CHECKSUM
```
enable it by default. Or why not?

maass-hamburg commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,786 @@
+	const struct device *clk_tx_dev;
+	uint8_t clk_tx_id;
+	const struct device *clk_rx_dev;
+	uint8_t clk_rx_id;
```
group the structs and the uint8_t together for a more efficient use

maass-hamburg commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,786 @@
+		res = k_sem_take(&data->rx_int_sem, K_FOREVER);
+		if (res == 0) {
+			while ((pkt = eth_rx(dev)) != NULL) {
+				iface = net_pkt_iface(pkt);
```
Can't you use the iface from the dev data struct and assign it before the while loop?

maass-hamburg commented on drivers/mdio/mdio_wch.c:
```diff
@@ -0,0 +1,158 @@
+	config->regs->MACMIIDR = data;
+	config->regs->MACMIIAR = tmpreg;
+
+	uint32_t read_start = k_uptime_get_32();
```
There is a timepoint* api use that.

maass-hamburg commented on dts/riscv/wch/ch32v307/ch32v307.dtsi:
```diff
@@ -198,6 +202,13 @@
+		rng: rng@40023c00 {
+			compatible = "wch,rng";
+			reg = <0x40023c00 12>;
+			clocks = <&rcc CH32V20X_V30X_CLOCK_RNG>;
```
Don't disable it, if it is added to the chosen.

maass-hamburg commented on dts/riscv/wch/ch32v303/ch32v303.dtsi:
```diff
@@ -198,6 +198,13 @@
+			compatible = "wch,rng";
+			reg = <0x40023c00 12>;
+			clocks = <&rcc CH32V20X_V30X_CLOCK_RNG>;
+			status = "disabled";
```
Should be enabled by default, as there are probably no external dependencies

Colt-Ma commented on dts/riscv/wch/ch32v307/ch32v307.dtsi:
The ch32v208 also has a 10M Ethernet interface and a built-in PHY. It would be great if this code could be adapted to the ch32v208.

BOJIT replied:
Hi @Colt-Ma ,
I'm afraid the CH32V203/208 Ethernet is a completely different peripheral.
The CH32F207/V307/V317 uses this synopsys IP: https://www.synopsys.com/designware-ip/interface-ip/ethernet/ether-mac10-100-universal.html (pretty common, is virtually identical to the STM32F4/F7 MAC)

The CH32F208/V203/V208 MCUs use a much more limited peripheral that is probably best managed as a separate driver rather than trying to make the same driver work for both.

maass-hamburg reviewed (changes requested):
> This has been fully tested on the CH32V307 dev board with the internal 10M PHY

I suppose that is the CH32V307V-EVT-R1?

Can you add that? Because I currently don't see there being a board in zephyr with the CH32V307 and without that the ethernet driver will not be build tested by the ci. Every ethernet driver has to be build tested by the ci.

BOJIT commented:
> > This has been fully tested on the CH32V307 dev board with the internal 10M PHY
> 
> I suppose that is the CH32V307V-EVT-R1?
> 
> Can you add that? Because I currently don't see there being a board in zephyr with the CH32V307 and without that the ethernet driver will not be build tested by the ci. Every ethernet driver has to be build tested by the ci.

I am using an SCDZ dev kit, virtually identical to the official EVT. The board was attempted to be added here by someone else (https://github.com/zephyrproject-rtos/zephyr/pull/90872) but the PR stalled.

I am happy to add a tidied version of this board, but can I request to get these two supporting PRs merged first?
- https://github.com/zephyrproject-rtos/zephyr/pull/101688
- https://github.com/zephyrproject-rtos/zephyr/pull/101391

Currently I'm already trying to juggle multiple unmerged interdependent patches and it's become a bit hard to manage.

loblik commented:
> I suppose that is the CH32V307V-EVT-R1?
> 
> Can you add that? Because I currently don't see there being a board in zephyr with the CH32V307

I just opened a PR for that one https://github.com/zephyrproject-rtos/zephyr/pull/103651 .

jpedley commented:
> > I suppose that is the CH32V307V-EVT-R1?
> > Can you add that? Because I currently don't see there being a board in zephyr with the CH32V307
> 
> I just opened a PR for that one #103651 .

@loblik would you mind testing Ethernet on this board if I add the DTS changes to your board PR?

Also FYI a lot of the Ethernet examples need the RAM/FLASH size configured larger to run on the WCH boards ([details here](https://blog.dan.drown.org/ch32v307-dev-board-part-5/))

I'm still wondering what the best way to handle this in Zephyr is: either board variants or perhaps snippets like [espressif](https://docs.zephyrproject.org/latest/boards/espressif/common/board-variants.html) uses? Not sure how these would integrate with the CI tests though.

loblik commented:
> @loblik would you mind testing Ethernet on this board if I add the DTS changes to your board PR?

Sure.

maass-hamburg commented:
> @maass-hamburg is this kind of thing accepatble?

I don't think so. Kconfigs can't be used in the dt. see https://docs.zephyrproject.org/latest/build/cmake/index.html#configuration-phase the dt is processed before the Kconfig.

BOJIT commented:
> Do I really need to do some pin configuration, when using internal 10Base PHY?

Yep, as the output pins of the PHY can apparently be remapped to do other things (think this is not the case on the V317 100M PHY though)

<img width="1065" height="129" alt="image" src="https://github.com/user-attachments/assets/b0a1cc5f-b5c9-4d48-98e7-1b51262e0634" />

My board support looks like this:

```dts
// Pinctrl
&pinctrl {
	mac_default: mac_default {
		mac_internal_rxp {
			pinmux = <CH32V20X_V30X_PINMUX_DEFINE(PC, 6, ETH, 0)>;
			slew-rate = "max-speed-50mhz";
		};

		mac_internal_rxn {
			pinmux = <CH32V20X_V30X_PINMUX_DEFINE(PC, 7, ETH, 0)>;
			slew-rate = "max-speed-50mhz";
		};

		mac_internal_txp {
			pinmux = <CH32V20X_V30X_PINMUX_DEFINE(PC, 8, ETH, 0)>;
			slew-rate = "max-speed-50mhz";
		};

		mac_internal_txn {
			pinmux = <CH32V20X_V30X_PINMUX_DEFINE(PC, 9, ETH, 0)>;
			slew-rate = "max-speed-50mhz";
		};
	};

	mdio_default: mdio_default {
		/* No MDIO pinmux required for internal PHY */
	};
};

// Main DTS
&mac {
	status = "okay";
	pinctrl-0 = <&mac_default>;
	pinctrl-names = "default";
	phy-connection-type = "internal";
	phy-handle = <&eth_phy>;
};

&mdio {
	status = "okay";
	pinctrl-0 = <&mdio_default>;
	pinctrl-names = "default";

	eth_phy: ethernet-phy@0 {
		compatible = "ethernet-phy";
		reg = <0x00>;
	};
};
```

Note that the MDIO driver treats the internal PHY like an external PHY. It still uses the same physical bus, it's just it's muxed to the internal PHY when it's enabled.

Hang tight! I'm planning to add ETH support to your board as part of this PR this weekend.

BOJIT commented:
@loblik I've added the additions to your board on the end of this PR.

The first commit adds the Ethernet peripheral to the board, the second adds variants for the different Flash/RAM splits on the SOC.

A lot of the network samples overflow the `32k` default RAM config. You can build this others with: `-b ch32v307v_evt_r1/ch32v307/ram_96k`, etc...

It doesn't seem ideal doing this at the board level, but that was the approach suggested to me here: https://github.com/zephyrproject-rtos/zephyr/issues/70721

BOJIT commented:
@maass-hamburg This should now have all the changes required. Unless @loblik has any concerns with the board additions I think this is ready for review.

maass-hamburg commented on boards/wch/ch32v307v_evt_r1/ch32v307v_evt_r1_defconfig:
```diff
@@ -6,3 +6,4 @@ CONFIG_GPIO=y
 CONFIG_SERIAL=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
+CONFIG_NET_L2_ETHERNET=y
```
not needed, already covered by the Kconfig.defconfig. Also we don't want the activated when we don't have Networking enabled.

maass-hamburg commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,764 @@
+			     DT_INST_ENUM_HAS_VALUE(inst, phy_connection_type, rmii) ||            \
+			     DT_INST_ENUM_HAS_VALUE(inst, phy_connection_type, rgmii) ||           \
+			     DT_INST_ENUM_HAS_VALUE(inst, phy_connection_type, internal),          \
+		     "Unsupported PHY connection type");                                           \
```
it seems like we only need to know, if the `phy_connection_type` is `internal` or not. Maybe leave that prop optional.

BOJIT committed:
boards: wch: add Ethernet to ch32v307_evt_r1

Adds enabled Ethernet to CH32V307 dev kit.
This ensures that the Ethernet driver is built in twister tests.

Signed-off-by: James Bennion-Pedley <james@bojit.org>

BOJIT committed:
drivers: ethernet: Support CH32V Ethernet Peripheral

This adds initial support for the CH32V ethernet peripheral.
The driver supports both internal and external PHY configurations.

Signed-off-by: James Bennion-Pedley <james@bojit.org>

BOJIT commented on drivers/ethernet/eth_wch.c:
```diff
@@ -0,0 +1,764 @@
+	tmpreg &= ~ETH_MACCR_FES_MASK;
+	tmpreg |= PHY_LINK_IS_SPEED_1000M(state->speed)  ? ETH_MACCR_FES_1000M
+		  : PHY_LINK_IS_SPEED_100M(state->speed) ? ETH_MACCR_FES_100M
+							 : ETH_MACCR_FES_10M;
```
Thanks for spotting, this got lost during a refactor!
Doesn't really show up as a problem when I test as I'm usually using PHY autonegotiation.

BOJIT committed:
boards: wch: add board variants for flash/ram split

The WCH Chips have the option to trade extra RAM for zero-waitstate Flash.
These configurations are added as board variants.

Signed-off-by: James Bennion-Pedley <james@bojit.org>

BOJIT commented:
@maass-hamburg if you get a chance could you take another look? Ideally trying to get this merged before starting some wch-platform-wide refactors.

BOJIT commented:
@pdgendt is it possible for you to have a look at this PR?
Sorry to keep pestering, but currently this PR is blocking https://github.com/zephyrproject-rtos/zephyr/pull/105534 along with a few other PTP-related features that I'm trying to get added to the WCH platform.

ceolin reviewed (approved):
+1 for the entropy changes.

I removed myself as the assignee because there are bigger changes than entropy in this pr.
